backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -224,4 +240,69 @@ private int getAvailableCores(List<VdsCpuUnit> cpuTopology, int socket, int vThr
+     * The cpuTopology is a deep copy made locally for this specific call, therefore we can change it.
+     * @param cpuTopology List<{@link VdsCpuUnit}>. The list of VdsCpuUnit.
+     * @param vm VM object.
+     * @param vdsId GUID of the VDS object.
+     */
+    private void filterBadNumaMemorySockets(List<VdsCpuUnit> cpuTopology, VM vm, Guid vdsId) {
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            return;
+        }
+
+        List<VdsNumaNode> vdsNumaNodes = vdsNumaNodeDao.getAllVdsNumaNodeByVdsId(vdsId);
+        if (vdsNumaNodes == null || vdsNumaNodes.isEmpty()) {
+            return;
+        }
+
+        Long highestMemRequired = getHighestNumaMemory(vm);
+        for (int socket : getOnlineSockets(cpuTopology)) {
+            List<Integer> numaIdsInSocket = getNumaIdsInSocket(cpuTopology, socket);
+            List<VdsNumaNode> numasInSocket = vdsNumaNodes.stream()
+                    .filter(numa -> numaIdsInSocket.contains(numa.getIndex()))
+                    .collect(Collectors.toList());
+            Long totalMemory = numasInSocket.stream()
+                    .map(VdsNumaNode::getNumaNodeStatistics)
+                    .map(NumaNodeStatistics::getMemFree)
+                    .reduce(0L, Long::sum);
+            if (totalMemory < highestMemRequired) {
+                // Memory is missing or there is no enough total memory available for the socket.
+                cpuTopology.removeAll(cpuTopology.stream().filter(cpu -> cpu.getSocket() == socket).collect(Collectors.toList()));
+            }
+        }
+    }
+
+    private List<Integer> getNumaIdsInSocket(List<VdsCpuUnit> cpuTopology, int socket) {
+        return cpuTopology.stream().filter(cpu -> cpu.getSocket() == socket).map(VdsCpuUnit::getNuma).distinct().collect(Collectors.toList());
+    }
+
+    private Long getHighestNumaMemory(VM vm) {
+        return Collections.max(vm.getvNumaNodeList().stream().map(NumaNode::getMemTotal).collect(Collectors.toList()));
+    }
+
+    private int getVirtualNumaNodesInSocket(List<VdsCpuUnit> cpuTopology, VM vm, Guid vdsId, int socket) {
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            return Integer.MAX_VALUE;
+        }
+
+        List<VdsNumaNode> vdsNumaNodes = vdsNumaNodeDao.getAllVdsNumaNodeByVdsId(vdsId);
+        if (vdsNumaNodes == null || vdsNumaNodes.isEmpty()) {
+            return Integer.MAX_VALUE;
+        }
+
+        List<Integer> numaIdsInSocket = getNumaIdsInSocket(cpuTopology, socket);
+        List<VdsNumaNode> numasInSocket = vdsNumaNodes.stream()
+                .filter(numa -> numaIdsInSocket.contains(numa.getIndex()))
+                .collect(Collectors.toList());
+        Long totalMemory = numasInSocket.stream()
+                .map(VdsNumaNode::getNumaNodeStatistics)
+                .map(NumaNodeStatistics::getMemFree)
+                .reduce(0L, Long::sum);
+        return (int) (totalMemory / getHighestNumaMemory(vm));
```

Comment:
just to understand this part, if I have a pnuma node with 4g
and I have vm with 3 nodes - 1g, 1g and 2g
this calculation would show that I can put at most 2 vnuma nodes in this socket although I can place 3?
it looks like we're ok with filtering sockets that may fit - do I get this correctly?

Reply:
yes, but i went with the thought they should be pretty close to each other in terms of memory.
the number of vcpu / vnuma count -> how many cpus within each vnuma.
memory is calculated by that (if memory can be divided without leftovers), or it will be set equal in memory.

the only case it's possible to have such difference in my opinion is only if the user changed it manually in the API.